### DISCUSSION TITLE
Transmittance Flux Region Depth Problem with Metal Grating

### DISCUSSION THREAD

**Discussion by QifengHu** [NanoComp/meep#2877] [Q&A] [answered]
Hi guys,

I am testing a simple 2D simulation of grating structure (Ag on SiO2) to investigate the reflectance and transmittance. I found that when electric field is parallel to the grating lines, the transmission flux region depth has a large impact on the short wavelength regimes. As I know, it should have no effect on the results.
![effect_of_TranFluxH2_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_200px](https://github.com/user-attachments/assets/d3eae438-4da9-411f-9816-41eaf39fa834)
![effect_of_TranFluxH4_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_200px](https://github.com/user-attachments/assets/5eecfe46-d526-4d20-b017-9d0797af3bce)
![effect_of_TranFluxH6_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_200px](https://github.com/user-attachments/assets/d90309a4-9675-4ac7-a486-ef36f614960d)
![effect_of_TranFluxH8_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_200px](https://github.com/user-attachments/assets/d317a979-6135-44ae-a47c-dc622bccaca6)
![effect_of_TranFluxH10_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_200px](https://github.com/user-attachments/assets/6f2837bc-608d-4d32-ab0b-2cf54fb9a50d)
![Reflectance_at_0_35_um_vs_Hf_T](https://github.com/user-attachments/assets/5d3130d0-f286-484c-a7b5-0df1fc386ad7)

Through the other tests, I confirmed that reflected region height has no influence on reflectance, but why transmittance at short wavelength (i.e., 0.35 μm) is different when the depth of the transmittance flux region height changes? I am really struggling with this issue. If you get a moment, please look at the following code I built.

Thank you very much.

Best,
Qifeng 

The code I built is look like this:
```py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Tue Jul 23 18:03:15 2024

@author: egrip-server-linux
"""

import meep as mp
import math
import numpy as np
import matplotlib.pyplot as plt
from meep.materials import SiO2, Ag
# valid wavelength range: (0.25, 1.77)
import pandas as pd

# Setup parameters are unchanged
resolution = 200
dpml = 1.0
dSiO2 = 10
gh = 0.03
gp = 0.5
dpad = 10

gdc = 0.9
s_height = 6
r_height = 2
t_depth = 2

sx = 2 * dpml + dSiO2 + gh + dpad
sy = gp

cell_size = mp.Vector3(sx, sy, 0)
pml_layers = [mp.PML(thickness=dpml, direction=mp.X)]

wvl_min = 0.3
wvl_max = 1.7
fmin = 1 / wvl_max
fmax = 1 / wvl_min

fcen = 0.5 * (fmax + fmin)
df = fmax - fmin
nfreq = 50

source_height = -sx / 2 + dpml + dSiO2 + 1 * gh + s_height
sources = [
    mp.Source(mp.GaussianSource(fcen, fwidth=df),
              component=mp.Ez,
              center=mp.Vector3(x=source_height),
              size=mp.Vector3(y=sy))
]

k_point = mp.Vector3(0, 0, 0)
symmetries = [mp.Mirror(mp.Y)]

# Initialize the simulation
sim = mp.Simulation(cell_size=cell_size,
                    boundary_layers=pml_layers,
                    sources=sources,
                    k_point=k_point,
                    symmetries=symmetries,
                    resolution=resolution)

# Add flux monitor for reflectance

#Creation of a FluxRegion centered at 
ref_height = -sx / 2 + dpml + dSiO2 + 1 * gh + r_height
refl_fr =  mp.FluxRegion(center=mp.Vector3(x=ref_height), size=mp.Vector3(y=sy))
refl_mon = sim.add_flux(fcen, df, nfreq, refl_fr)

trans_height =  -sx / 2 + dpml + dSiO2 + gh - t_depth
trans_fr = mp.FluxRegion(center=mp.Vector3(x=trans_height), size=mp.Vector3(y=sy))
trans_mon = sim.add_flux(fcen, df, nfreq, trans_fr)

# Run the simulation
sim.run(until_after_sources=mp.stop_when_fields_decayed(10, mp.Ez, mp.Vector3(x=trans_height), 1e-9))

# Calculate reflected flux
empty_flux = mp.get_fluxes(refl_mon)
empty_data = sim.get_flux_data(refl_mon) # store the Flux Data from the EMPTY RUN into "empty_data"

#empty_trans_flux = mp.get_fluxes(trans_mon)
#empty_trans_data = sim.get_flux_data(trans_mon)

sim.reset_meep()


# Redefine the geometry including any layers, blocks, etc.
geometry = [
    mp.Block(
        mp.Vector3(dpml + dSiO2, mp.inf, mp.inf),
        center=mp.Vector3(-sx / 2 + 0.5* dpml + 0.5 * dSiO2),
        material=SiO2
    ),
    mp.Block(
        mp.Vector3(gh, gdc * gp, mp.inf),
        center=mp.Vector3(-sx / 2 + dpml + dSiO2 + 0.5 * gh),
        material=Ag
    )
]

# Reinitialize the simulation with the same parameters but updated geometry
sim = mp.Simulation(
    resolution=resolution,
    cell_size=cell_size,
    boundary_layers=pml_layers,
    geometry=geometry,
    k_point=k_point,
    sources=sources,
    symmetries=symmetries
)

#adding the FluxRegion "refl_fr" to the simulation
refl = sim.add_flux(fcen, df, nfreq, refl_fr)
trans = sim.add_flux(fcen, df, nfreq, trans_fr)

#Load the EMPTY RUN Flux Data using .load_minus_flux_data to SUBTRACT the fluxes from the next run.
# (Basically removing the empty run fluxes to leave only the reflectance)
sim.load_minus_flux_data(refl, empty_data)
# sim.load_minus_flux_data(trans, empty_trans_data)


# Run the simulation
sim.run(#mp.at_beginning(mp.output_epsilon), 
        until_after_sources=mp.stop_when_fields_decayed(10, mp.Ez, mp.Vector3(x=trans_height), 1e-9))


refl_flux = mp.get_fluxes(refl) # store the flux data of the Second run in "refl_flux"
R_meep = -1*np.divide(refl_flux,empty_flux) #calculate the reflectance from MEEP simulation runs

trans_flux = mp.get_fluxes(trans)
T_meep = np.divide(trans_flux, empty_flux)
# Reset the simulation to reuse the structure and source definition
sim.reset_meep()







freqs = mp.get_flux_freqs(refl)
wvls = np.divide(1,freqs)

# post_process
if mp.am_really_master() == True:
    legend_text = [
        f"Thickness of SiO2: {dSiO2}μm",
        f"Thickness of Ag: {gh}μm",
        f"Grating Period: {gp}μm",
        "Space ratio: {}".format(gdc),
        "Resolution: {} pixels/μm;".format(resolution),
        "Source Height: {} μm".format(s_height),
        "Flux Region Height of Reflectance: {} μm".format(r_height),
        "Flux Region Depth of Transmittance: {} μm".format(t_depth),
        f'Wavelength range: [{wvl_min}, {wvl_max}]',
        f'Number of frequency: {nfreq}',
    ]

    # Combine into a single string with newline characters
    legend_label = '\n'.join(legend_text)

    df = pd.DataFrame({"Wavelength": wvls, "Reflectance": R_meep, "Transimittance": T_meep})
    df.to_csv(f"data/wvls{wvl_min}_{wvl_max}_n{nfreq}_R_T_ratio{gdc}_dSiO2{dSiO2}_SourceH{s_height}_FluxRegionH_R{r_height}_T{t_depth}_{resolution}px.csv")

    # Create the figure and the first axis
    fig, ax = plt.subplots()
    ax.plot(wvls,R_meep,'bo-', label = 'Reflectance')
    ax.plot(wvls, T_meep, 'ro-', label = "Transmittance")
    ax.plot(wvls, R_meep+T_meep, 'k--', label='R+T')
    ax.set_xlabel("Wavelength (μm)")
    ax.set_ylabel("Power Percentage")
    ax.legend(frameon=False)
    ax.set_ylim([0.,1.])
    
    fig.set_size_inches(9,4)
    plt.figtext(0.7, 0.2, legend_label, wrap=True, horizontalalignment='left', fontsize=12)
    plt.subplots_adjust(right=0.65)
    
        # Show the plot
    plt.title("Flux Region Depth of Transmittance: {} μm".format(t_depth))
    plt.savefig(f"pic/effect_of_TranFluxH{t_depth}_EPara2Line_spaceratio{gdc}_SourceH{s_height}_ReflFluxH{r_height}_{resolution}px.png")
    plt.show()
```

**Comment by stevengj** [comment]
It would be easier to help you if you made a diagram of your simulation via `sim.plot2D()`, since I'm not sure what "depth" you are talking about.

(Also, be sure check convergence with resolution.)

**Reply by QifengHu** [reply]
Thank you for your help, I attached the diagram of the structure. 
![geometry_visualization](https://github.com/user-attachments/assets/714fcee7-3b08-4482-8de3-327d7a71c21c)

The "depth" means the distance between measured flux region for transmittance and the top surface of metal (Ag). 
Also, I attached a resolution convergence study on one problem setting: depth=2 microns. The transmittance and reflectance keep unchanged when resolution increases to 100 px/μm. Thus we selected resolution of 200 px/ μm in all previous simulations. The conditions are the same as above; with 'space ratio' meaning Metal:Dielectric width ratio = 0.9 and the basic geometry shown in the above figure. It seems to show below that the convergence is pretty good after we go past the very coarse region at the left. We are requiring convergence test at the end of the simulation of 10^-9 so we think we are indeed getting convergence at all test regions. 
We also have problems with the long-wave reflection of this grating since the reflection for E perp to the lines should be low, as expected for a wire grid polariser, but in fact is large; and transmittance was very low. In addition we wonder about the silica thickness since when we go to a 300um thickness of the silica underlayer, and vary the flux region below the Ag grating, we find that >12 micron flux region depth we get weirdly unreasonable transmittance results at long wavelength. We have those results and could open another GitHub ticket for those two other issues ... unless we make this a longer ticket. 

![Reflectance_at_0_35_um_vs_Resolution](https://github.com/user-attachments/assets/e79d2840-dc36-45c2-a757-a315b7ef2577)

**Comment by mochen4** [comment, marked as answer]
I noticed that you used the predefined "SiO2" material, which, I think, includes material loss. That would explain why changing the location of transmittance monitor would have different result. Instead, if you define a lossless material, e.g. mp.Medium(index=1.55), the transmittance should be the same.

**Reply by QifengHu** [reply]
Thank you, Mo. It's indeed the reason. Now I got the consistent results when changing monitor locations of transmittance. The power percentage is shown as followed:
![silica_effect_of_refraidx1 55_EPara2Line_spaceratio0 9_SourceH6_ReflFluxH2_TranFluxH2_200px](https://github.com/user-attachments/assets/c23cdc6c-9656-4597-a284-a3a430320268)
The other thing we wonder is why there is a high transmittance peak at short wavelength. It is not the SiO2 absorption causing a transmittance peak.

**Reply by mochen4** [reply]
It could be due to the Ag grating (which is also lossy: if you additionally specify a lossless material instead of the pre-defined "Ag", then R+T should identically equal to one). Depending on the material/geometry/wavelength, there may be some resonant mode at certain wavelength causing the peak in transmittance.
